File: services/server/src/safe_socket/safe_socket.py
import socket

def recv_all(socket: socket.socket, size):
    chunks = []
    bytes_recd = 0
    while bytes_recd < size:
        chunk = socket.recv(size - bytes_recd)
        if not chunk:
            raise RuntimeError("Socket connection broken")
        chunks.append(chunk)
        bytes_recd += len(chunk)
    return b''.join(chunks)


def send_all(socket: socket.socket, bytes):
    total_sent = 0
    while total_sent < len(bytes):
        sent = socket.send(bytes[total_sent:])
        if sent == 0:
            raise RuntimeError("Socket connection broken")
        total_sent += sent

File: services/client/src/safe_socket/safe_socket.go
package safe_socket

import "io"

func SendAll(socket io.Writer, bytes []byte) error {
	for totalSent := 0; totalSent < len(bytes); {
		n, err := socket.Write(bytes[totalSent:])
		if err != nil {
			return err
		}
		totalSent += n
	}
	return nil
}

func RecvAll(socket io.Reader, size int) ([]byte, error) {
	buff := make([]byte, size)
	for recv := 0; recv < size; {
		n, err := socket.Read(buff[recv:])
		recv += n
		if err != nil {
			return buff[:recv], err
		}
	}
	return buff, nil
}
